### ISSUE TITLE
twisted.conch.ssh.userauth depends on hash ordering

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5776]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)| @pitrou reported|
|-|-|
|Trac ID|trac#5776|
|Type|defect|
|Created|2012-07-13 17:02:44Z|

Running the cftp script can fail with hash randomization:

```
$ python2.6 -R bin/conch/cftp -v pitrou.net
2012-07-13 18:58:43+0200 [-] Log opened.
2012-07-13 18:58:43+0200 [-] Starting factory <twisted.conch.client.direct.SSHClientFactory instance at 0x2118f80>
2012-07-13 18:58:43+0200 [SSHClientTransport,client] kex alg, key alg: diffie-hellman-group-exchange-sha1 ssh-rsa
2012-07-13 18:58:43+0200 [SSHClientTransport,client] outgoing: aes256-ctr hmac-sha1 none
2012-07-13 18:58:43+0200 [SSHClientTransport,client] incoming: aes256-ctr hmac-sha1 none
2012-07-13 18:58:43+0200 [SSHClientTransport,client] REVERSE
2012-07-13 18:58:43+0200 [SSHClientTransport,client] NEW KEYS
2012-07-13 18:58:43+0200 [SSHClientTransport,client] setting client server to <twisted.conch.client.default.SSHUserAuthClient instance at 0x2118ef0>
2012-07-13 18:58:43+0200 [SSHClientTransport,client] starting service ssh-userauth
2012-07-13 18:58:43+0200 [SSHClientTransport,client] using agent
2012-07-13 18:58:43+0200 [SSHAgentClient,client] got 3 public keys
2012-07-13 18:58:43+0200 [SSHService ssh-userauth on SSHClientTransport,client] can continue with: ['publickey', 'password']
2012-07-13 18:58:43+0200 [SSHService ssh-userauth on SSHClientTransport,client] trying to auth with publickey
2012-07-13 18:58:43+0200 [SSHService ssh-userauth on SSHClientTransport,client] using key of type RSA
2012-07-13 18:58:43+0200 [SSHService ssh-userauth on SSHClientTransport,client] can continue with: ['publickey', 'password']
2012-07-13 18:58:43+0200 [SSHService ssh-userauth on SSHClientTransport,client] trying to auth with publickey
2012-07-13 18:58:43+0200 [SSHService ssh-userauth on SSHClientTransport,client] using key of type DSA
2012-07-13 18:58:43+0200 [SSHService ssh-userauth on SSHClientTransport,client] can continue with: ['publickey', 'password']
2012-07-13 18:58:43+0200 [SSHService ssh-userauth on SSHClientTransport,client] trying to auth with publickey
2012-07-13 18:58:43+0200 [SSHService ssh-userauth on SSHClientTransport,client] using key of type DSA
2012-07-13 18:58:43+0200 [SSHService ssh-userauth on SSHClientTransport,client] couldn't handle 60
[snip]
```

At this point, the script blocks instead of connecting successfully.
That's because in userauth.py, the messages table is initialized from the globals() dictionary, and there are duplicate message values. Message number 60 can be mapped to MSG_USERAUTH_PASSWD_CHANGEREQ, MSG_USERAUTH_INFO_REQUEST or MSG_USERAUTH_PK_OK depending on ordering. Without hash randomization, it is always mapped to MSG_USERAUTH_PK_OK.

Attachments:

* [randomhash_userauth.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a59/a59414da705d122842c95ca7d7770728a3cba967/373860622f289a67447a090fc1f26b9d94289893.patch) (1294 bytes) - added by antoine on 2012-07-13 17:10:35Z - 
* [randomhash_userauth2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a59/a59414da705d122842c95ca7d7770728a3cba967/a6c6e254011403fe696185bbf9ddf793a806d47b.patch) (1946 bytes) - added by antoine on 2012-07-14 12:51:28Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5776 5776
type__defect defect
reporter__antoine antoine
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__None None
time__1342198964000000 1342198964000000
changetime__1474197555542286 1474197555542286
version__None None
owner__ 
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_2"></a>|@pitrou commented|
|-|-|

Patch attached.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_3"></a>|@mwhudson set owner to @pitrou|
|-|-|

Hi, thanks for fixing the problem and for adding the testcase.  The behaviour around here is pretty horrible :/

Can I ask you to, instead of sorting the names, simply assign MSG_USERAUTH_PASSWD_CHANGEREQ and 
MSG_USERAUTH_INFO_REQUEST after the loop?

Also can you add a news file per http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles ?  I'll leave it up to you as to whether this is 'misc' or 'bugfix' :-)

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_4"></a>|@pitrou commented|
|-|-|

Replying to [mwh](#note_3):
> 
> Can I ask you to, instead of sorting the names, simply assign MSG_USERAUTH_PASSWD_CHANGEREQ and 
> MSG_USERAUTH_INFO_REQUEST after the loop?

That's a better idea indeed. I'm attaching a new patch.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)|@pitrou removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_7"></a>|@mwhudson set status to `closed`|
|-|-|

(In [34749]) Merge randomhash_userauth2.patch: twisted.conch.ssh.userauth now works correctly with hash randomization enabled.

Author: antoine
Reviewer: mwh
Fixes: [#5776](https://github.com/twisted/twisted/issues/5776)

The userauth code depended on mapping message type 60 to MSG_USERAUTH_PK_OK but
that only happened because of luck with dictionary ordering.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing review from closed tickets.
